### ISSUE TITLE
getTerreinOp() returns value from array based on non zero-based value

### ISSUE THREAD

**Issue by TheBauwssss** [gameoftrades/gameoftrades-student-kit#3]
Function `getTerreinOp(Coordinaat c)` in the file `Kaart.java` on line number `35`:

``` java
public Terrein getTerreinOp(Coordinaat c) {
    if ((c.getX() < 0) || (c.getY() < 0) || (c.getX() >= getBreedte()) || (c.getY() >= getHoogte())) {
        throw new IllegalArgumentException(c + " ligt buiten de Kaart.");
    }
    return terrein[c.getY()][c.getX()];
}
```

The array `terrein` is zero based, but the `getX()` and `getY()` from `Coordinaat` return values starting at 1, thus returning the wrong `Terrein` for any given `Coordinate.`

**Comment by timvisee** [comment]
I think the terrain tile being accessed (whether that's zero- or one-based) depends on the way the `Coordinate` class is used inside your project, and isn't necessarily defined by the function above as you seem to suggest.

When creating a `Coordinate` instance, you should use zero-based values too, instead of one-based values. You might currently be creating one-based `Coordinate` instances in your project, while the top-left tile is at `(0, 0)` and not `(1, 1)` as specified [here](https://github.com/gameoftrades/gameoftrades-student-kit/blob/master/OPDRACHT.md#tips), even though the latter might seem more logical.

The function you've mentioned above simply uses the raw coordinate (on the X and Y axis), and doesn't convert the actual values between zero- and one-based schemes. All other functions seem to act in the same way.

Are you indeed sure you're using zero-based values everywhere when using `Coordinate`'s, because I can't reproduce the same problem?
Or am I not seeing the real problem you're mentioning?

**TL;DR:** Make sure you're using the zero-based coordinate system throughout your whole project, don't mix them up.

**Comment by TheBauwssss** [comment]
The instances used when reproducing the bug are not created by me. They are passed from the `SnelstePadDebugPanel`. 

The function that calculates the shortest path between two coordinates gets passed a `Kaart` instance and two `Coordinaat` instances. These `Coordinaat` instances are supposed to represent to two selected `Stad` instances between which a path needs to be calculated. But, when `kaart.getCoordinaatOp()` is invoked on any of those coordinates the returned 'Terrein' instance is not the `Stad` instance, but in fact the instance at x+1 and y+1 relative to the `Stad` instance.

**Comment by TheBauwssss** [comment]
As you can see in this screenshot, `kaart.getTerreinOp()` only returns the correct `Stad` instances after translating the coordinates by -1.

![](http://rick-soft.com/sharex_uploads/2016-09/idea_2016-09-20_20-02-37_2.png)

**Comment by timvisee** [comment]
The _begin_ and _end_ coordinates are directly fetched from the `Stad` classes, as you can see [here](https://github.com/gameoftrades/gameoftrades-student-kit/blob/master/src/test/java/io/gameoftrades/studentNN/HandelaarImplTest.java#L56-L63) in the test included with the student kit:
![Coordinates directly fetched from cities](http://i.imgur.com/p66jZf6.png)

Are you sure you're instantiating the `Stad` classes with zero-based coordinates?

The coordinates stored in (and fetched from) `Stad` classes do not seem to be translated in the library itself, ever, unless I'm missing something. In that case, instantiating them with one-based coordinates would result in the problem you describe above.

**Comment by TheBauwssss** [comment]
Yes, I see. The `Stad` instances were indeed being instantiated with one based values. Weird that all tests were green while this is obviously a bug.

**Comment by timvisee** [comment]
I've created a small manual test suite to test the problem you're suggesting. Check out the branch [here](https://github.com/timvisee/got-team49-public/tree/coord-test).

Manual test coverage:
- The map tiles, along with cities and other relevant objects are loaded from a map file.
- The test algorithm and cities are selected in the interface (by the user).
- The `SnelstePadAlgoritme` implementation is invoked through the interface.
- The terrain type tiles of the start and end coordinates (without translations) are printed to the console, which can be validated by the user.

_Note: The following test is dirty, and should only be used to verify the coordinate system is working correctly described by the test case._

## 

This test includes a simple (dummy) fastest past algorithm class called `FastPathAlgorithmTest` which implements the `SnelstePadAlgoritme` interface. When this algorithm class is invoked it prints the type of tile associated with the start and end coordinates to the console, which were passed to the `FastPathAlgorithmTest#bereken` method when it is invoked.

I've added two test cities to make things easy, these cities are called `Test City 1` and `Test City 2` and can be found in the `westeros-kaart.txt` map.

You can run application with the test by starting the `StudentUI` class (debug mode not required). Please select the `FastPathAlgorithmTest` algorithm when the main interface comes up. Also make sure you select `Test City 1` and `Test City 2` as `van` and `naar` cities. Then, click the `Start` button to start the test algorithm. This should print the terrain type of the start and end coordinates to the console as I've mentioned earlier.

I get the following output when running this test:
![Test output](http://i.imgur.com/12NPVLq.png)

Facts:
- The city `Test City 1` is located at `(0, 0)` which is a `ZEE` tile in the map configuration.
- The city `Test City 2` is located at `(1, 1)` which is a `BERG` tile in the map configuration.

The result shown in the picture above correspond to these facts, and thus give correct results.
Based on this test, I think it's safe to say you don't have to translate any coordinates to get correct results.

## 

**Update:** It seems you've figured out what was causing the problem, awesome!
This is indeed untested behavior, but I don't agree it's a _bug_.
I'll leave the comment here because it might be of help for other users running into a similar problem and/or for reference.
